lib/fluent/plugin/out_mongo.rb

```diff
@@ -213,23 +214,30 @@ def collect_records(chunk)
         if date_keys
           date_keys.each { |date_key|
             begin
-              date_value = record[date_key]
+              date_key_accessor = record_accessor_create(date_key)
+              date_value = date_key_accessor.call(record)
+              if date_value.to_i.to_s == date_value
+                date_value = date_value.to_i
+              elsif date_value.to_f.to_s == date_value
+                date_value = date_value.to_f
+              end
               case date_value
               when Fluent::EventTime
-                record[date_key] = date_value.to_time
+                value_to_set = date_value.to_time
               when Integer
-                record[date_key] = if date_value > 9999999999
+                value_to_set = if date_value > 9999999999
                                      # epoch with milliseconds: e.g. javascript
                                      Time.at(date_value / 1000.0)
                                    else
                                      # epoch with seconds: e.g. ruby
                                      Time.at(date_value)
                                    end
               when Float
-                record[date_key] = Time.at(date_value)
+                value_to_set = Time.at(date_value)
               else
-                record[date_key] = Time.parse(date_value)
+                value_to_set = Time.parse(date_value)
               end
+              date_key_accessor.set(record, value_to_set)
```

Comment:
Any chance to add testcase(s) for this change?
We encourage to do unittesting when adding/updating functionality.

Reply:
I have implemented two more unittest (one valid and one invalid) as you proposed